app/controllers/dashboard_controller.rb

```diff
@@ -14,6 +14,34 @@ def index
     redirect_to :action => 'show'
   end
 
+  def _cockpit_url
+    # Behind apache use paths only
+    url = URI(MiqCockpit::ApacheConfig.url_root)
+
+    # No apache, use current host unless external url configured
+    unless request.env['HTTP_X_FORWARDED_FOR']
+      hostname = URI.parse(request.original_url).hostname
+      conf = MiqCockpitWsWorker.worker_settings
+      url = MiqCockpit::WS.url(conf || {}, hostname)
+    end
+    url
+  end
+
+  # Redirect to cockpit with an api auth token
+  def cockpit_redirect
+    url = _cockpit_url
+
+    # If there is a redirect uri, use only the path from there.
+    if params[:redirect_uri]
+      url.path = URI.parse(params[:redirect_uri]).path
+    elsif params[:ip]
+      url.path = url.path + "=" + params[:ip] + "/"
```

Comment:
If you have a URI object, you shouldn't need to do string manipulation.  (nor in the fragment below)